cmd/crank/render/render.go

```diff
@@ -187,18 +187,52 @@ func Render(ctx context.Context, log logging.Logger, in Inputs) (Outputs, error)
 	runner := composite.NewFetchingFunctionRunner(runtimes, &FilteringFetcher{extra: in.ExtraResources})
 
 	observed := composite.ComposedResourceStates{}
+	secretGvk := schema.GroupVersionKind{
+		Group:   "",
+		Version: "v1",
+		Kind:    "Secret",
+	}
+
+	observedConnectionDetails := map[xpv1.SecretReference]managed.ConnectionDetails{}
+
+	for _, cd := range in.ObservedResources {
+		if secretGvk != cd.GroupVersionKind() {
+			continue
+		}
+		secret := &corev1.Secret{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(cd.Object, secret); err != nil {
+			continue
+		}
+		observedConnectionDetails[xpv1.SecretReference{
+			Name:      secret.GetName(),
+			Namespace: secret.GetNamespace(),
+		}] = secret.Data
+	}
+
 	for i, cd := range in.ObservedResources {
-		name := cd.GetAnnotations()[AnnotationKeyCompositionResourceName]
+		name, ok := cd.GetAnnotations()[AnnotationKeyCompositionResourceName]
+		if !ok {
+			continue
+		}
+		var or resource.Composed = &in.ObservedResources[i]
+		var connectionDetails managed.ConnectionDetails
+		if connectionSecretRef := or.GetWriteConnectionSecretToReference(); connectionSecretRef != nil {
+			connectionDetails = observedConnectionDetails[*connectionSecretRef]
+		}
```

Comment:
I could but I think it would be a little awkward since i cannot assign directly to a [field of a map](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#UnaddressableFieldAssign).

I would have to do something like this:
```go
if ref := or.GetWriteConnectionSecretToReference(); ref != nil {
	crs := observed[composite.ResourceName(name)]
	crs.ConnectionDetails = observedConnectionDetails[*ref]
	observed[composite.ResourceName(name)] = crs
}
```

but i tried to improve the spacing and variable names to make it easier to read.